File: frontend/gameapp-frontend/src/domain/value-objects/enemies/enemy.js
import { EnemyName } from './enemy-name.js';

export class Enemy {
    static ERROR_INVALID_NAME = "Parameter 'name' must be an instance of EnemyName";
    static ERROR_INVALID_NUMBER = (param) => `Parameter '${param}' must be a finite number`;
    static ERROR_INVALID_DAMAGE = "Damage must be a finite number";

    constructor(name, healthPoints, attackDamage, speedAttack, rewardMoney) {
        this.#validateName(name);
        this.#validateNumber(healthPoints, "healthPoints");
        this.#validateNumber(attackDamage, "attackDamage");
        this.#validateNumber(speedAttack, "speedAttack");
        this.#validateNumber(rewardMoney, "rewardMoney");

        this._name = name;
        this._healthPoints = healthPoints;
        this._attackDamage = attackDamage;
        this._speedAttack = speedAttack;
        this._rewardMoney = rewardMoney;
    }

    #validateName(name) {
        if (!(name instanceof EnemyName)) {
            throw new TypeError(Enemy.ERROR_INVALID_NAME);
        }
    }

    #validateNumber(value, paramName) {
        if (typeof value !== "number" || !Number.isFinite(value)) {
            throw new TypeError(Enemy.ERROR_INVALID_NUMBER(paramName));
        }
    }

    // Getters
    get name() { return this._name; }
    get healthPoints() { return this._healthPoints; }
    get attackDamage() { return this._attackDamage; }
    get speedAttack() { return this._speedAttack; }
    get rewardMoney() { return this._rewardMoney; }

    // Setters
    setName(newName) {
        return new Enemy(newName, this._healthPoints, this._attackDamage, this._speedAttack, this._rewardMoney);
    }

    setHealthPoints(newHealthPoints) {
        return new Enemy(this._name, newHealthPoints, this._attackDamage, this._speedAttack, this._rewardMoney);
    }

    setAttackDamage(newAttackDamage) {
        return new Enemy(this._name, this._healthPoints, newAttackDamage, this._speedAttack, this._rewardMoney);
    }

    setSpeedAttack(newSpeedAttack) {
        return new Enemy(this._name, this._healthPoints, this._attackDamage, newSpeedAttack, this._rewardMoney);
    }

    setRewardMoney(newRewardMoney) {
        return new Enemy(this._name, this._healthPoints, this._attackDamage, this._speedAttack, newRewardMoney);
    }

    receiveDamage(damage) {
        this.#validateNumber(damage, "damage");

        let newHealthPoints = this._healthPoints - damage;
        if (newHealthPoints < 0) newHealthPoints = 0;

        return new Enemy(this._name, newHealthPoints, this._attackDamage, this._speedAttack, this._rewardMoney);
    }

    toString() {
        return `${this._name.toString()} Enemy: ` +
            `HealthPoints=${this._healthPoints}, ` +
            `AttackDamage=${this._attackDamage}, ` +
            `SpeedAttack=${this._speedAttack}, ` +
            `RewardMoney=${this._rewardMoney}`;
    }

}

File: backend/GameApp.Domain/Entities/Enemy.cs
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Combat;
using GameApp.Domain.ValueObjects.Enemies;

namespace GameApp.Domain.Entities;

// Class for an enemy
public class Enemy
{
    private readonly Guid Id;
    private readonly EnemyName Name;
    private readonly int HealthPoints;
    private readonly int AttackDamage;
    private readonly int SpeedAttack;
    private readonly int RewardMoney;
    private readonly CriticalDamage CriticalDamage;
    private readonly EnemyDifficulty Difficulty;


    // Default constructor
    public Enemy(EnemyDifficulty difficulty, EnemyName name, int healthPoints, int attackDamage, int speedAttack, CriticalDamage criticalDamage, int rewardMoney)
    {
        Id = Guid.NewGuid();
        Difficulty = difficulty;
        Name = name;
        HealthPoints = healthPoints;
        AttackDamage = attackDamage;
        SpeedAttack = speedAttack;
        CriticalDamage = criticalDamage;
        RewardMoney = rewardMoney;
    }

    // Restore constructor
    public Enemy(Guid id, EnemyDifficulty difficulty, EnemyName name, int healthPoints, int attackDamage, int speedAttack, CriticalDamage criticalDamage, int rewardMoney)
    {
        Id = id;
        Difficulty = difficulty;
        Name = name;
        HealthPoints = healthPoints;
        AttackDamage = attackDamage;
        SpeedAttack = speedAttack;
        CriticalDamage = criticalDamage;
        RewardMoney = rewardMoney;
    }

    // Getters
    public Guid GetGuid() => Id;
    public EnemyDifficulty GetDifficulty() => Difficulty;
    public EnemyName GetName() => Name;
    public int GetHealthPoints() => HealthPoints;
    public int GetAttackDamage() => AttackDamage;
    public int GetSpeedAttack() => SpeedAttack;
    public int GetRewardMoney() => RewardMoney;
    public CriticalDamage GetCriticalDamage() => CriticalDamage;
    

    // Setters
    public Enemy SetDifficulty(EnemyDifficulty newDifficulty) => new Enemy(GetGuid(), newDifficulty, Name, HealthPoints, AttackDamage, SpeedAttack, CriticalDamage, RewardMoney);
    public Enemy SetName(EnemyName newName) => new Enemy(GetGuid(), Difficulty, newName, HealthPoints, AttackDamage, SpeedAttack, CriticalDamage, RewardMoney);
    public Enemy SetHealthPoints(int newHealthPoints) => new Enemy(GetGuid(), Difficulty, Name, newHealthPoints, AttackDamage, SpeedAttack, CriticalDamage, RewardMoney);
    public Enemy SetAttackDamage(int newAttackDamage) => new Enemy(GetGuid(), Difficulty, Name, HealthPoints, newAttackDamage, SpeedAttack, CriticalDamage, RewardMoney);
    public Enemy SetSpeedAttack(int newSpeedAttack) => new Enemy(GetGuid(), Difficulty, Name, HealthPoints, AttackDamage, newSpeedAttack, CriticalDamage, RewardMoney);
    public Enemy SetRewardMoney(int newRewardMoney) => new Enemy(GetGuid(), Difficulty, Name, HealthPoints, AttackDamage, SpeedAttack, CriticalDamage, newRewardMoney);
    public Enemy SetCriticalDamage(CriticalDamage newCriticalDamage) => new Enemy(GetGuid(), Difficulty, Name, HealthPoints, AttackDamage, SpeedAttack, newCriticalDamage, RewardMoney);
    public Enemy ReceiveDamage(int damage)
    {
        int newHealthPoints = HealthPoints - damage;

        if (newHealthPoints < 0)
        {
            newHealthPoints = 0;
        }

        return new Enemy(Difficulty, Name, newHealthPoints, AttackDamage, SpeedAttack, CriticalDamage, RewardMoney);
    }

    // To string
    public override string ToString()
    {
        return $"{Name.GetName()} Enemy(Difficulty {Difficulty}): " +
               $"HealthPoints={HealthPoints}, " +
               $"AttackDamage={AttackDamage}, " +
               $"SpeedAttack={SpeedAttack}, " +
               $"Critical Damage={CriticalDamage}" +
               $"RewardMoney={RewardMoney}";
    }

}